core/amber/src/main/scala/edu/uci/ics/amber/engine/common/promise/PromiseManager.scala

```diff
@@ -0,0 +1,221 @@
+    tryInvokeNextSyncPromise()
+  }
+
+  // send a control message to another actor, and keep the handle.
+  def schedule[T](cmd: PromiseBody[T], on: ActorVirtualIdentity = selfID): Promise[T] = {
+    val ctx = mkPromiseContext()
+    promiseID += 1
+    controlOutputPort.sendTo(on, PromiseInvocation(ctx, cmd))
+    val promise = WorkflowPromise[T](promiseContext)
+    unCompletedPromises(ctx) = promise
+    promise
+  }
+
+  // send a grouped control message to other actors, and keep the handle.
+  def schedule[T](seq: (PromiseBody[T], ActorVirtualIdentity)*): Promise[Seq[T]] = {
+    val promise = WorkflowPromise[Seq[T]](promiseContext)
+    if (seq.isEmpty) {
+      // if the sequence is empty, resolve the promise immediately
+      promise.setValue(Seq.empty)
+    } else {
+      unCompletedGroupPromises.add(
+        GroupedWorkflowPromise[T](promiseID, promiseID + seq.length, promise)
+      )
+      seq.foreach {
+        case (body, virtualIdentity) =>
+          val ctx = mkPromiseContext()
+          promiseID += 1
+          controlOutputPort.sendTo(virtualIdentity, PromiseInvocation(ctx, body))
+      }
+    }
+    promise
+  }
+
+  def returning(value: Any): Unit = {
+    // returning should be used at most once per context
+    if (promiseContext != null) {
+      controlOutputPort.sendTo(promiseContext.sender, ReturnPayload(promiseContext, value))
+      exitCurrentPromise()
+    }
+  }
+
+  def returning(): Unit = {
+    // returning should be used at most once per context
+    if (promiseContext != null) {
+      controlOutputPort.sendTo(
+        promiseContext.sender,
+        ReturnPayload(promiseContext, PromiseCompleted())
+      )
+      exitCurrentPromise()
+    }
+  }
+
+  @inline
+  private def exitCurrentPromise(): Unit = {
+    if (ongoingSyncPromises.contains(promiseContext)) {
+      ongoingSyncPromises.remove(promiseContext)
+    }
+    promiseContext = null
+  }
+
+  @inline
+  private def tryInvokeNextSyncPromise(): Unit = {
+    if (ongoingSyncPromises.isEmpty) {
+      syncPromiseRoot = null
+      if (queuedInvocations.nonEmpty) {
+        execute(queuedInvocations.dequeue())
+      }
+    }
+  }
+
+  @inline
+  private def registerSyncPromise(rootCtx: RootPromiseContext, ctx: PromiseContext): Unit = {
+    syncPromiseRoot = rootCtx
+    ongoingSyncPromises.add(ctx)
+  }
+
+  @inline
+  protected def mkPromiseContext(): PromiseContext = {
+    promiseContext match {
+      case null =>
+        // if current context is null, create root context
+        PromiseContext(selfID, promiseID)
+      case ctx: RootPromiseContext =>
+        // create child context with root
+        PromiseContext(selfID, promiseID, ctx)
+      case ctx: ChildPromiseContext =>
+        // use ctx.root to create child context
+        PromiseContext(selfID, promiseID, ctx.root)
+    }
+  }
+
+  @inline
+  private def invokePromise(ctx: PromiseContext, call: PromiseBody[_]): Unit = {
+    promiseContext = ctx
+    promiseHandler(call)
```

Comment:
We are not checking what type of control message it is?

Reply:
promiseHandler is a PartialFunction[Any, Unit], basically, it is a Receive in actors. And we write different handlers for different types of control messages.